### ISSUE TITLE
Selective merge of branches

### ISSUE THREAD

**Issue by yohami** [playcanvas/editor#654]
Following up with the idea on #340 https://github.com/playcanvas/editor/issues/340 for handling a shared framework across different games -

Currently we have a game per branch and the shared framework on a separate branch.

The request is to be able to merge in and out of the framework branch without also sending / deleting the game assets and scripts from each branch.

Our current methodology is to pull all games in the framework branch, merge with the specific game branch, delete all other games there, work, then pull from master and commit again to master. Works but it's less than ideal and the shared framework branch is turning unmanegeable with all the game assets from all branches there

**Comment by Sproilex** [comment]
I agree, it's really needed when trying to work on projects with multiple subprojects based on a main Framework. Being able to select wich files and folders we want to merge (Like in most of the source control systems out there) would be a big improvement regarding to velocity and risk.

**Comment by yaustar** [comment]
> Being able to select wich files and folders we want to merge (Like in most of the source control systems out there) 

I'm not aware of this being possible with other source control where everything is in the same repo. Unless you mean something like cherry picking?

**Comment by yohami** [comment]
The shared framework is all contained in a folder - it has scripts and assets inside. Then each game has a folder for it's own scripts and assets (plus scenes, entities on screen etc). Ideally we could merge in and out of the assets folder without touching the game folders.

So, cherry picking but it doesnt have to be file by file, could be folder based.

**Comment by yaustar** [comment]
Sounds similar to sub repos in this case? I think the linking of projects is more inline of the request than source control merging?

**Comment by yohami** [comment]
True sounds like a sub repo.

Linking of projects could work as well - the thing is that we develop the main framework as we work on features for a specific game. So the main framework is changing constantly depending on the features we add to a game, and the idea is that such features then can be reused by all games. 

I think that with linking of projects we wouldn't be able to edit the linked project while working on another? but if that's possible then it serves the same function

**Comment by Sproilex** [comment]
> I'm not aware of this being possible with other source control systems where everything is in the same repo. Unless you mean something like git cherry picking?

There are tricky ways to get a similar selective merging function on other source control systems (I normally use Git but i hear about similar functionality in others), in git for example cherry picking can be one but there's also a command to do it directly, using `git merge --no-ff --no-commit branch` you will be able to then remove files that you don't want to commit on the merge while keeping them in your local repo.

**Comment by yohami** [comment]
Good find @Sproilex

**Comment by yohami** [comment]
And yes something like that would solve our problems

**Comment by yaustar** [comment]
The problem with that method in the stackoverflow is that it only works in one direction:

> **BIG CAUTION. This merge works in one direction**, but the files that you did not include in the merge are seen as DELETED if you decide to merge your upstream master back into the source branch. I use a git-flow like process using a master branch (production main-line), a staging branch (staging server main-line), and topic branches based off of the staging branch. Using this strategy has resulted in my "reverse merge" from master back into staging to competely fail thinking everything I didn't merge from staging to master is deleted. This includes whole files and hunks. YOU ARE WARNED

**Comment by yaustar** [comment]
That said, we do have a method to ignore textual changes during merges https://github.com/playcanvas/playcanvas-sync#the-pcignoretxt-file

I wonder if that extends or could be extended to work with assets and scenes?

**Comment by yohami** [comment]
pcignoretext - I had forgotten about it but yes that sounds like the easiest way so far. We would just need to be able to write a folder name there and have git ignore it's whole contents

I'll check with Leonidas if we can hack this with a script so the pcignore is automatically filled, will be in touch :-)

**Comment by yaustar** [comment]
I will also check on my side about the extent of it and whether it could be used for scenes as well or have that be added

**Comment by LeXXik** [comment]
I would think that adding pcignore.txt that would list folders to ignore would act the same way as I would use .gitignore specifying a folder to ignore. However, this doesn't seem to be the case. If a branch has pcignore.txt in its root, then it doesn't matter what it specifies (to be ignored?) nothing is ignored and the whole branch is merged to master. Parhaps the functionality between pcignore and gitignore is not the same.

**Comment by yaustar** [comment]
I need to check with @zachpeterpaul on this.

IIRC, if you merge to a branch that has pcignore.txt, it will ignore changes to the files listed in the pcignore.txt via the syntax in the reader above

**Comment by zachpeterpaul** [comment]
Please remember to take a checkpoint after adding a pcignore.txt file (to the merge destination branch) and before your merge. Also (from the readme): Use a space and not * or ? to match a space in a file or folder name in gitignore lines. This is because of some limitations of the third party libraries used to parse that file.

**Comment by yohami** [comment]
@zachpeterpaul how would you add a folder to the pcignore?

We tried foldername/*

**Comment by zachpeterpaul** [comment]
please try /path/foldrename
no slash or star at the end.
Also, pcignore won't exclude files from the result, it just keeps them the same before and after merge.

**Comment by yohami** [comment]
Thanks -

The case is that the branch has a folder and we want to avoid sending it to the branch we're merging into. Is that possible?

**Comment by zachpeterpaul** [comment]
No that's not possible

**Comment by yohami** [comment]
Then that's our feature request :-)

**Comment by yohami** [comment]
What if we list all of the assets in the folder, would that prevent them to be sent to the branch we're merging into?

**Comment by zachpeterpaul** [comment]
No, that won't help. You can create a new branch from your source branch, delete those folders there, and merge from that branch.

**Comment by yohami** [comment]
That's what we're doing but it's impractical, at some point the master branch will crash due to having all the games in it

We have a master branch with the shared framework. Then a branch per game. Each game branch has all of it's assets and scripts - but in the process of merging in and out the shared framework, master then has all of the assets and scripts for all games, and those can't be deleted from master or they also get deleted when you pull master into a game

So we're looking for a solution for this

**Comment by yohami** [comment]
> delete those folders there, and merge from that branch.

If you do this, then when you merge from the branch it also deletes the folders on the branch you're merging into


1) Create master branch
2) Create another branch for a game, pull all game assets there
3) Work
4) Merge into master
5) Now master has all of the unwanted game assets, delete folder
6) Do some work on master / shared framework
7) Merge master into game branch -> this deletes all of the game assets on that branch

We want to be able to exclude the game assets folder from the merging process

**Comment by zachpeterpaul** [comment]
I'll look into this, since this affects our merge it won't be a quick change. pcignore is for different types of use cases. I'm afraid if one simply excludes files from merge completely, there will be no way to collaborate on them via different branches/merges. I think the way git handles this, is to allow a subfolder referencing another git repo (e.g. with your common code/shared framework). Then merging the outer folder (your individual game code) can ignore the subfolder. The subfolder would have to have a whole copy of the referenced repo, with the ability to pull into it, mapping all our internal id's to the current (outer) project. What do you think? I can discuss this further with our team and consider adding something like this to our roadmap.

**Comment by zachpeterpaul** [comment]
continued: It wouldn't have to be restricted to just a subfolder of assets, whole scenes or entity subtrees could be treated as external, not affected by merges, and only updatable via a special external pull operation.

**Comment by yohami** [comment]
Actually git version (gitignore) does allow to skip a whole folder from the repo

Having two projects and loading one into another is an alternative, we were discussing it here https://github.com/playcanvas/editor/issues/340 

That can work well as long as we're able to edit the assets of one project from another? the shared framework is developed as we work on specific games. It's not a static import

**Comment by zachpeterpaul** [comment]
I think adding some syntax to pcignore to skip folders is possible. Re shared code/frameworks, yes one would be able to update it to the latest version via a special pull operation. Each shared item would have a "standard" id in its own project/repo, and projects that use it will have a map from standard id's to their local copies. Our templates work in a similar way.

**Comment by yohami** [comment]
Both approaches are promising, let me know what the team says :-)

**Comment by zachpeterpaul** [comment]
Sounds good!

**Comment by yaustar** [comment]
Reposting from email that I'm having with one of your team members

From the current feature set of PlayCanvas Editor, I would be doing the following (note that I'm not invalidating the feature request, just outlining my approach to this):

Single project, one game per branch approach

The shared SDK would be in the master branch and then there would be a branch per game. Changes to the SDK would be made against the master branch and then it can be merged upstream to the game branches as needed.

Each branch then could be configured as needed without the 'weight' of the other games. The version control graph view and careful naming of branches should help with management.

To migrate from the current state of one game per scene to one game per branch, this is the way I would do:
- Make all the branches for the games
- On master delete all the game files so it's just the framework/shared code
- Merge master to the game branches (which will delete the game files on the branches)
- On the game branch, restore the first checkpoint to get back the game files and makes a new checkpoint on the game branch that is ahead of the master deletion of the game files.

The downside of this is that SDK changes can only be merged upstream (master to the game branches) unless pcignore.txt can be used to 'block'/ignore changes to the game assets and files/folders from the game.

The master branch that is responsible for the SDK would have a 'shell game', a very thin game that is there to purely test the SDK. Eg. it would have a button on screen to play an ad rather than actual game logic. Any changes to the SDK should be made on master or a branch off master with the intention to merge back to master when done.

This workflow keeps each game branch small and there is a clear workflow for making changes and testing the SDK. Each game branch can also control when to take the latest version of the SDK on master by choosing when to merge up.

---

There is a risk concern with your current workflow and with the above suggestion as all your work is still in a single project.

With that, I do recommend regular backups to archive the project and necessary branches via the REST API https://developer.playcanvas.com/en/user-manual/api/project-archive/ regardless which workflow you continue using.

We will discuss internally on what we can do to properly support a shared SDK workflow and see if we can make changes in the short term as well to help.

**Comment by yohami** [comment]
Thanks @yaustar, we'll give it a try

**Comment by yaustar** [comment]
We have some ideas and will be going through a design and review stage for a feature that sounds more like #340 than selective merging

**Comment by yohami** [comment]
Cool
